activation/nipost_state.go

```diff
@@ -20,43 +19,64 @@ const (
 	challengeFilename = "nipost_challenge.bin"
 	builderFilename   = "nipost_builder_state.bin"
 	postFilename      = "post.bin"
-	crc64Size         = 8
 )
 
 func write(path string, data []byte) error {
-	buf := bytes.NewBuffer(nil)
+	tmp, err := os.CreateTemp("", filepath.Base(path))
+	if err != nil {
+		return fmt.Errorf("create temporary file %s: %w", tmp.Name(), err)
+	}
+
+	crc := make([]byte, crc64.Size)
+	if _, err = tmp.Write(crc); err != nil { // Add padding for the CRC at the start of the file.
+		return fmt.Errorf("write checksum padding %s: %w", tmp.Name(), err)
+	}
+
 	checksum := crc64.New(crc64.MakeTable(crc64.ISO))
-	w := io.MultiWriter(buf, checksum)
-	if _, err := w.Write(data); err != nil {
-		return fmt.Errorf("write data %v: %w", path, err)
+	w := io.MultiWriter(tmp, checksum)
+	if _, err = w.Write(data); err != nil {
+		return fmt.Errorf("write data %v: %w", tmp.Name(), err)
 	}
 
-	crc := make([]byte, crc64Size)
 	binary.BigEndian.PutUint64(crc, checksum.Sum64())
-	if _, err := buf.Write(crc); err != nil {
-		return fmt.Errorf("write checksum %v: %w", path, err)
+	if _, err = tmp.WriteAt(crc, 0); err != nil {
```

Comment:
is it a common practice to append crc at the start? what are the disadvantages of adding it as a suffix? i think crc as a suffix allows to preserve sequential reads, it obviously doesn't matter here

Reply:
The [first point of the specification on Wikipedia](https://en.wikipedia.org/wiki/Cyclic_redundancy_check) seems to indicate it is. To be taken with a grain of salt.

That being said I would argue having some fixed length data first and the variable length data afterwards makes for a better design. especially if the variable length data is really large. It is also the way it was specified in the issue.

Also looking further into it, adding a zeros prefix does not change the CRC so I can rewrite the `write` function to remove the initial write of the prefix.

OR I can rewrite it with the CRC as a suffix. The logic might be a bit more complicated for the read because we need the logic to stop before the end of the file but that can work too.